modules/packages/ParallelIO.chpl

```diff
@@ -0,0 +1,852 @@
+              r = locFile.reader(locking=false, region=taskBounds, deserializer=des);
+
+        var item: t, first = true;
+        for i in tItemOffsets[tid]..<tItemOffsets[tid+1] {
+          if first then first = false; else r.advanceThrough(delim);
+          r.read(item);
+          results[i] = item;
+        }
+      }
+    }
+
+    return results;
+  }
+
+  /*
+    Read a delimited file in parallel into an array.
+
+    This procedure is essentially the same as :proc:`readDelimitedAsBlockArray`,
+    except that it only executes on the calling locale. As such, it does not
+    accept a ``targetLocales`` argument and returns a non-distributed array.
+
+    :arg filePath: a path to the file to read from
+    :arg delim: the delimiter to use to separate ``t`` values in the file
+    :arg t: the type of value to read from the file
+    :arg nTasks: the number of tasks to use
+    :arg header: how to handle the file header (see :record:`headerPolicy`)
+    :arg deserializerType: the type of deserializer to use
+
+    :returns: a default rectangular array of ``t`` values
+
+    :throws: :class:`OffsetNotFoundError` if a starting offset cannot be found in
+              any of the chunks
+
+    See :proc:`~IO.open` for other errors that could be thrown when attempting
+    to open the file
+  */
+  proc readDelimitedAsArray(filePath: string, type t, in delim: ?dt = b"\n",
+                            header = headerPolicy.noHeader, nTasks: int = here.maxTaskPar,
+                            type deserializerType = defaultDeserializer
+  ): [] t throws
+    where dt == bytes || dt == string
+  {
+    const f = open(filePath, ioMode.r),
+          fileBounds = 0..<f.size,
+          byteOffsets = findDelimChunks(f, delim, nTasks, fileBounds, header),
+          itemOffsets = findItemOffsets(f, delim, byteOffsets);
+
+    const resDom = {0..<itemOffsets.last};
+    var results: [resDom] t;
+
+    coforall tid in 0..<nTasks with (ref results) {
+      var des: deserializerType;
+      const taskBounds = byteOffsets[tid]..<byteOffsets[tid+1],
+            r = f.reader(locking=false, region=taskBounds, deserializer=des);
+
+      var item: t, first = true;
+      for i in itemOffsets[tid]..<itemOffsets[tid+1] {
+        if first then first = false; else r.advanceThrough(delim);
+        r.read(item);
+        results[i] = item;
+      }
+    }
+
+    return results;
+  }
+
+  /*
+    Read items from a file in parallel into a block-distributed array.
+
+    This routine assumes that the file is composed of a series of deserializable
+    values of type ``t`` (optionally with a header at the beginning of the file).
+    Each ``t`` must be separated by a delimiter which can either be provided as
+    a ``string`` or ``bytes`` value. Unlike :proc:`readDelimitedAsBlockArray`
+    the delimiter can also be found in the serialized form of ``t``.
+
+    This routine uses the following heuristic to split the file into chunks:
```

Comment:
Can it say more directly somewhere here that this heuristic is not 100% accurate in all cases?